java/core/src/main/java/co/worklytics/psoxy/gateway/output/ApiDataOutputUtils.java

```diff
@@ -0,0 +1,267 @@
+package co.worklytics.psoxy.gateway.output;
+
+import co.worklytics.psoxy.ControlHeader;
+import co.worklytics.psoxy.gateway.HttpEventRequest;
+import co.worklytics.psoxy.gateway.ProcessedContent;
+import com.google.api.client.http.HttpRequest;
+import com.google.api.client.http.HttpResponse;
+import lombok.AllArgsConstructor;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.http.HttpHeaders;
+
+import javax.inject.Inject;
+import javax.inject.Provider;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URLDecoder;
+import java.nio.charset.StandardCharsets;
+import java.util.*;
+import java.util.stream.Collectors;
+
+@AllArgsConstructor(onConstructor_ = {@Inject})
+public class ApiDataOutputUtils {
+
+    /**
+     * keys for metadata that will be added to API Data output objects.
+     */
+    public enum OutputObjectMetadata {
+
+        /**
+         * the host to which the request was sent, eg, api.saas-tool.com
+         */
+        API_HOST,
+
+        /**
+         * the HTTP method of the request, eg, GET, POST, ...
+         */
+        HTTP_METHOD,
+
+        /**
+         * the path of the request
+         */
+        PATH,
+
+        /**
+         * query string of the request, if any
+         */
+        QUERY_STRING,
+
+        /**
+         * base64-encoded body of the request, if any (eg, for POST requests)
+         */
+        REQUEST_BODY,
+        ;
+    }
+
+    Provider<UUID> uuidProvider;
+    Base64.Encoder base64encoder;
+
+
+    public String buildRawOutputKey(HttpRequest requestToSourceApi) {
+        Map<String, String> metadata = buildRawMetadata(requestToSourceApi);
+        return buildRawOutputKey(metadata);
```

Comment:
I'd put a limit on keys for 1024 characters. We can do 992 characters or regular request value and it exceeds, put the rest into MD5 hashed (32 more characters)

Reply:
i've restructured this so it's just `yyyyMMdd/{uuid}`.  makes it easy to inspect that things are "working" by looking in the bucket, still concise.  Avoids need to worry about sanitization of the path.  Also don't need to worry about unuseful repetition of host + path components, which aren't adding any value.